fastddsspy_participants/src/cpp/visualization/ModelParser.cpp

```diff
@@ -214,65 +214,219 @@ ComplexParticipantData ModelParser::participants(const SpyModel& model, const dd
+        {
+            result.push_back(readers(model, it.second.guid));
+        }
+    }
+
     return result;
 }
 
 ComplexEndpointData ModelParser::readers(const SpyModel& model, const ddspipe::core::types::Guid& guid)
 {
     ComplexEndpointData result;
-    // TODO
+
+    for (const auto& it : model.endpoint_database_)
+    {
+        if (it.first == guid && it.second.is_reader())
+        {
+            result = {
+                it.second.guid,
+                it.second.discoverer_participant_id ,
+                {
+                    it.second.topic.m_topic_name,
+                    it.second.topic.type_name
+                },
+                {
+                    std::to_string(it.second.topic.topic_qos.durability_qos),
+                    std::to_string(it.second.topic.topic_qos.reliability_qos)
+                }
+            };
+            break;
+        }
+    }
+
     return result;
 }
 
 
 std::vector<SimpleTopicData> ModelParser::topics(const SpyModel& model)
 {
     std::vector<SimpleTopicData> result;
-    // TODO
+    std::map<std::string, int> already_topics_added;
+
+    for (const auto& it1 : model.endpoint_database_)
+    {
+        int datawriters = 0;
+        int datareaders = 0;
+        auto repeat = already_topics_added.find(it1.second.topic.m_topic_name);
+        if (repeat == already_topics_added.end())
+        {
+            already_topics_added[it1.second.topic.m_topic_name] = result.size();
+            for (const auto& it2 : model.endpoint_database_)
+            {
+                if (it2.second.topic.m_topic_name == it1.second.topic.m_topic_name)
+                {
+                    if (it2.second.is_reader())
+                    {
+                        datareaders++;
+                    }
+                    if (it2.second.is_writer())
+                    {
+                        datawriters++;
+                    }
+                }
+            }
+            result.push_back({
+                        it1.second.topic.m_topic_name,
+                        it1.second.topic.type_name,
+                        datawriters,
+                        datareaders,
+                        {
+                            10,     // not implementated yet
+                            "Hz"
+                        }
+                    });
+        }
+    }
+
     return result;
 }
 
 std::vector<ComplexTopicData> ModelParser::topics_verbose(const SpyModel& model)
 {
     std::vector<ComplexTopicData> result;
-    // TODO
+
+    for (const auto& it : model.types_discovered_)
+    {
+        result.push_back(topics(model, it.first));
+    }
+
     return result;
 }
 
 ComplexTopicData ModelParser::topics(const SpyModel& model, const std::string& topic_name)
 {
     ComplexTopicData result;
-    // TODO
+    std::vector<ComplexTopicData::Endpoint> datareaders;
+    std::vector<ComplexTopicData::Endpoint> datawriters;
+
+    std::string type_name = "Not discovered yet";
```

Comment:
In case the topic does not exist (look for it in endpoints database) return empty name